Adding text on running GEOPM runtime alongside non-MPI applications (#3108)
geopm/geopm: service/docs/source/runtime.rst

=======================

The GEOPM Runtime is software designed to enhance energy efficiency of

> applications though active hardware configuration.  

Reviewer: Very minor detail:

```suggestion
applications through active hardware configuration.  
```

User Model
----------

The architecture is

> designed to provide a secure infrastructure to support a wide range

Reviewer: Very minor nit: the line break is too early. Try to follow same column limit being followed in the rest of the document.

         trace file per host, and one report across all hosts.
   :align: center

.. admonition:: Quick start for MPI applications

   The ``geopmlaunch`` tool is the recommended user interface to launch the GEOPM Runtime for 

>    profiling and optimizing MPI applications. It

Reviewer: Minor nit: maybe review column limits, follow same limit as in the rest of the document.

   wraps a launcher application (srun in this example), generates a summarizing
   report file, and optionally generates a time-series trace per host.

   The steps for using ``geopmlaunch`` with your MPI application are:

  * Specify how many nodes and processes to use.
  * Run the ``geopmlaunch`` command wherever you would normally run the
    wrapped launcher command (e.g., ``srun``, ``mpiexec``, etc.).
  * Read the generated ``geopm.report`` file
  
  .. code-block:: console

>     :caption: examples using ``geopmlaunch``

Reviewer 2: Captions should start with a capital letter in all the style guidelines I can dig up.  Also, the other use of ":caption:" in our docs follows this style.

    $ # Launch with srun and explore the generated geopm report
    $ geopmlaunch srun -N 1 -n 20 -- ./my-app
    $ less geopm.report
    $ # Launch with Intel mpiexec and explore the generated GEOPM report

>     $ geopmlaunch impi -N 1 -ppn 20 -- ./my-app

Reviewer 2: `mpiexec.hydra` which backs the impi launcher, does not support the `-N` argument.  This was correct the way it was.

```
$ geopmlaunch impi -N 1 -ppn 1 hostname
...
[mpiexec@mcfly10] match_arg (../../../../../src/pm/i_hydra/libhydra/arg/hydra_arg.c:91): unrecognized argument N
...
```

>     $ # launch with intel mpiexec and explore the generated geopm report

Reviewer 3: These lowercase changes look unintentional

    $ less geopm.report
    $ # show all options and available launchers
    $ geopmlaunch --help

.. admonition:: Quick start for Non-MPI applications

>    In order to profile non-MPI application, the recommended approach is to launch

Reviewer: nit:
```suggestion
   In order to profile non-MPI applications, the recommended approach is to launch
```

   the GEOPM runtime alongside the target application. This can be done by launching
   the ``geopmctl`` application in the background on every host node that the non-MPI 

>    application is expected to run on. After a clean or forced termination of the 

Reviewer: ```suggestion
   application is expected to run. After a clean or forced termination of the 
```

   application being profiled, the ``geopmctl`` application is designed to terminate
   and generate a summarizing report file, and optionally, a time-series trace per host. 
 
   The following requirements must be met while launching ``geopmctl`` with your 
   non-MPI application:
 
   * Both the ``geopmctl`` process and the application process must have
     the ``GEOPM_PROFILE`` environment variable set to the **same**
     value.
   * The application process must have ``LD_PRELOAD=libgeopm.so.1`` set
     in the environment, or the application binary must be linked

>      directly to ``libgeopm.so.1`` at compile time.

Reviewer 2: ```suggestion
   * The application process must have ``LD_PRELOAD=libgeopm.so.2`` set
     in the environment, or the application binary must be linked
     directly to ``libgeopm.so.2`` at compile time.
```

The ABI change went in last Friday.

>    * The ``GEOPM_REPORT`` environment variable must be set in the

Reviewer: I'm thinking it would also be helpful to mention the `GEOPM_PROGRAM_FILTER` environment variable (not as a requirement, but as an optional feature that could help the user ensure the right application is profiled).

Author: Added text & code snippets to runtime.rst, overview.rst, and geopm.7.rst

     environment of the ``geopmctl`` process.
 
   .. code-block:: console
     :caption: examples using ``geopmctl``
 
     $ GEOPM_PROFILE=sleep-ten \
       GEOPM_REPORT=sleep-ten.yaml \

>        geopmctl &

Reviewer 2: ```suggestion
     $ GEOPM_PROFILE=sleep-ten \
       GEOPM_REPORT=sleep-ten.yaml \
       GEOPM_TRACE=sleep-ten-trace \
       geopmctl &
```

     $ GEOPM_PROFILE=sleep-ten \

>        LD_PRELOAD=libgeopm.so.1 \

Reviewer 2: ```suggestion
       LD_PRELOAD=libgeopm.so.2 \
```

       sleep 10
     $ cat sleep-ten.yaml

>      $ awk -F\| '{print $1, $6, $8}' sleep-ten.csv | less

Reviewer 2: ```suggestion
     $ awk -F\| '{print $1, $6, $8}' sleep-ten-trace* | less
```

Remember that traces will be prepended with the hostname.

Please test all these examples by hand to ensure they're working as expected.

The `geopm runtime tutorial
<https://github.com/geopm/geopm/tree/dev/tutorial#geopm-tutorial>`_ shows how
to profile unmodified applications, select and evaluate different geopm agent

> algorithms, and how to add markup to an application.  the tutorial provides a

Reviewer: ```suggestion
algorithms, and how to add markup to an application.  The tutorial provides a
```

> starting point for someone trying to get familiar with the geopm runtime.

Reviewer 2: ```suggestion
starting point for someone trying to get familiar with the GEOPM runtime.
```

Unless you're referring to one of the command-line tools, or compiled files (libraries and the like), or source files keep GEOPM capitalized please.  We do this elsewhere on this page.

   the GEOPM runtime alongside the target application. This can be done by launching
   the ``geopmctl`` application in the background on every host node that the non-MPI 
   application is expected to run. After a clean or forced termination of the 
   application being profiled, the ``geopmctl`` application is designed to terminate
   and generate a summarizing report file, and optionally, a time-series trace per host. 
 
   The following requirements must be met while launching ``geopmctl`` with your 
   non-MPI application:
 
   * Both the ``geopmctl`` process and the application process must have
     the ``GEOPM_PROFILE`` environment variable set to the **same**
     value.
   * The application process must have ``LD_PRELOAD=libgeopm.so.2`` set
     in the environment, or the application binary must be linked
     directly to ``libgeopm.so.2`` at compile time.
   * The ``GEOPM_REPORT`` environment variable must be set in the
     environment of the ``geopmctl`` process.
   * The ``GEOPM_PROGRAM_FILTER`` environment variable can optionally be set to
     explicitly list the program invocation name of the non-MPI processes that need

>      to be profiled.

Reviewer: Maybe add a note mentioning why this option is useful (since it's optional), to avoid profiling any extraneous processes, or maybe adding a note similar to the one in the `geopmlaunch` documentation.
